File: internal/bot/factory.go
package bot

import (
	"fmt"
	"math/rand"
	"time"
)

// EngineOption is a functional option for engine creation.
type EngineOption func(*engineConfig) error

// engineConfig holds configuration options for engine creation.
type engineConfig struct {
	difficulty  Difficulty
	timeLimit   time.Duration
	searchDepth int
	options     map[string]any
}

// WithTimeLimit sets a custom time limit for move selection.
func WithTimeLimit(d time.Duration) EngineOption {
	return func(c *engineConfig) error {
		if d <= 0 {
			return fmt.Errorf("time limit must be positive")
		}
		c.timeLimit = d
		return nil
	}
}

// WithSearchDepth sets a custom search depth for minimax engines.
func WithSearchDepth(depth int) EngineOption {
	return func(c *engineConfig) error {
		if depth < 1 || depth > 20 {
			return fmt.Errorf("search depth must be 1-20")
		}
		c.searchDepth = depth
		return nil
	}
}

// WithOptions sets custom options as a map.
func WithOptions(opts map[string]any) EngineOption {
	return func(c *engineConfig) error {
		c.options = opts
		return nil
	}
}

// NewRandomEngine creates an Easy bot with random move selection.
func NewRandomEngine(opts ...EngineOption) (Engine, error) {
	cfg := &engineConfig{
		difficulty: Easy,
		timeLimit:  2 * time.Second,
	}

	for _, opt := range opts {
		if err := opt(cfg); err != nil {
			return nil, err
		}
	}

	return &randomEngine{
		name:      "Easy Bot",
		timeLimit: cfg.timeLimit,
		closed:    0, // atomic: 0 = open
		rng:       rand.New(rand.NewSource(time.Now().UnixNano())),
	}, nil
}

// NewMinimaxEngine creates a Medium or Hard bot using minimax with alpha-beta pruning.
func NewMinimaxEngine(difficulty Difficulty, opts ...EngineOption) (Engine, error) {
	cfg := &engineConfig{difficulty: difficulty}

	// Set defaults based on difficulty
	switch difficulty {
	case Medium:
		cfg.timeLimit = 4 * time.Second
		cfg.searchDepth = 4
	case Hard:
		cfg.timeLimit = 8 * time.Second
		cfg.searchDepth = 6
	default:
		return nil, fmt.Errorf("invalid difficulty for minimax: %d (expected Medium or Hard)", difficulty)
	}

	// Apply custom options
	for _, opt := range opts {
		if err := opt(cfg); err != nil {
			return nil, err
		}
	}

	// Create the minimax engine
	name := fmt.Sprintf("%s Bot", difficulty.String())

	return &minimaxEngine{
		name:        name,
		difficulty:  cfg.difficulty,
		maxDepth:    cfg.searchDepth,
		timeLimit:   cfg.timeLimit,
		evalWeights: getDefaultWeights(cfg.difficulty),
		closed:      false,
	}, nil
}

File: rust/crates/bot/src/factory.rs
//! Engine construction and functional options (ported from `factory.go`).

use std::collections::HashMap;
use std::time::Duration;

use crate::error::EngineError;
use crate::interfaces::Difficulty;
use crate::minimax::{get_default_weights, MinimaxEngine};
use crate::random::RandomEngine;

/// A value stored in the free-form options map (`map[string]any` in Go).
#[derive(Debug, Clone, PartialEq)]
pub enum OptionValue {
    /// An integer option value.
    Int(i64),
    /// A boolean option value.
    Bool(bool),
    /// A string option value.
    Str(String),
}

/// Configuration accumulated by [`EngineOption`]s during engine creation.
///
/// The struct is public so options can be passed to the factory functions, but
/// its fields are crate-private (matching Go's unexported `engineConfig`).
#[derive(Debug, Clone, PartialEq)]
pub struct EngineConfig {
    pub(crate) difficulty: Difficulty,
    pub(crate) time_limit: Duration,
    pub(crate) search_depth: i32,
    pub(crate) deterministic: bool,
    pub(crate) options: Option<HashMap<String, OptionValue>>,
}

impl Default for EngineConfig {
    fn default() -> Self {
        EngineConfig {
            difficulty: Difficulty::Easy,
            time_limit: Duration::ZERO,
            search_depth: 0,
            deterministic: false,
            options: None,
        }
    }
}

/// A functional option applied to an [`EngineConfig`] during engine creation.
pub type EngineOption = Box<dyn Fn(&mut EngineConfig) -> Result<(), EngineError>>;

/// Sets a custom time limit for move selection. Must be positive.
///
/// Note: Rust's `Duration` is unsigned, so only a zero limit is invalid here
/// (Go additionally rejected negative durations).
pub fn with_time_limit(d: Duration) -> EngineOption {
    Box::new(move |c| {
        if d.is_zero() {
            return Err(EngineError::Config(
                "time limit must be positive".to_string(),
            ));
        }
        c.time_limit = d;
        Ok(())
    })
}

/// Sets a custom search depth for minimax engines (1-20).
pub fn with_search_depth(depth: i32) -> EngineOption {
    Box::new(move |c| {
        if !(1..=20).contains(&depth) {
            return Err(EngineError::Config("search depth must be 1-20".to_string()));
        }
        c.search_depth = depth;
        Ok(())
    })
}

/// Sets custom options as a map.
pub fn with_options(opts: HashMap<String, OptionValue>) -> EngineOption {
    Box::new(move |c| {
        c.options = Some(opts.clone());
        Ok(())
    })
}

/// Disables random tie-breaking for reproducible results (used in tests).
pub fn with_deterministic(deterministic: bool) -> EngineOption {
    Box::new(move |c| {
        c.deterministic = deterministic;
        Ok(())
    })
}

/// Creates an Easy bot with weighted random move selection.
pub fn new_random_engine(opts: &[EngineOption]) -> Result<RandomEngine, EngineError> {
    let mut cfg = EngineConfig {
        difficulty: Difficulty::Easy,
        time_limit: Duration::from_secs(2),
        ..EngineConfig::default()
    };

    for opt in opts {
        opt(&mut cfg)?;
    }

    Ok(RandomEngine::new("Easy Bot", cfg.time_limit))
}

/// Creates a Medium or Hard bot using minimax with alpha-beta pruning.
pub fn new_minimax_engine(
    difficulty: Difficulty,
    opts: &[EngineOption],
) -> Result<MinimaxEngine, EngineError> {
    let mut cfg = EngineConfig {
        difficulty,
        ..EngineConfig::default()
    };

    match difficulty {
        Difficulty::Medium => {
            cfg.time_limit = Duration::from_secs(4);
            cfg.search_depth = 4;
        }
        Difficulty::Hard => {
            cfg.time_limit = Duration::from_secs(8);
            cfg.search_depth = 7;
        }
        _ => {
            return Err(EngineError::Config(format!(
                "invalid difficulty for minimax: {} (expected Medium or Hard)",
                difficulty as i32
            )));
        }
    }

    for opt in opts {
        opt(&mut cfg)?;
    }

    let name = format!("{} Bot", difficulty);

    Ok(MinimaxEngine::new(
        name,
        cfg.difficulty,
        cfg.search_depth,
        cfg.time_limit,
        get_default_weights(cfg.difficulty),
        cfg.deterministic,
    ))
}
